src/check_cache.cc

```diff
@@ -76,58 +74,57 @@ CheckCache::~CheckCache() {
 }
 
 void CheckCache::Check(const Attributes& attributes, CheckResult* result) {
-  std::string signature;
-  Status status = Check(attributes, system_clock::now(), &signature);
+  Status status = Check(attributes, system_clock::now());
   if (status.error_code() != Code::NOT_FOUND) {
     result->status_ = status;
   }
 
-  result->on_response_ = [this, signature](
-      const Status& status, const CheckResponse& response) -> Status {
+  result->on_response_ = [this](const Status& status,
+                                const Attributes& attributes,
+                                const CheckResponse& response) -> Status {
     if (!status.ok()) {
       if (options_.network_fail_open) {
         return Status::OK;
       } else {
         return status;
       }
     } else {
-      return CacheResponse(signature, response, system_clock::now());
+      return CacheResponse(attributes, response, system_clock::now());
     }
   };
 }
 
-Status CheckCache::Check(const Attributes& attributes, Tick time_now,
-                         std::string* ret_signature) {
+Status CheckCache::Check(const Attributes& attributes, Tick time_now) {
   if (!cache_) {
     // By returning NOT_FOUND, caller will send request to server.
     return Status(Code::NOT_FOUND, "");
   }
 
-  std::string signature = GenerateSignature(attributes, *cache_keys_);
-  if (ret_signature) {
-    *ret_signature = signature;
-  }
-
-  std::lock_guard<std::mutex> lock(cache_mutex_);
-  CheckLRUCache::ScopedLookup lookup(cache_.get(), signature);
+  for (const auto& it : referenced_map_) {
+    const Referenced& reference = it.second;
+    std::string signature;
+    if (!reference.Signature(attributes, &signature)) {
+      continue;
+    }
 
-  if (!lookup.Found()) {
-    // By returning NO_FOUND, caller will send request to server.
-    return Status(Code::NOT_FOUND, "");
+    std::lock_guard<std::mutex> lock(cache_mutex_);
+    CheckLRUCache::ScopedLookup lookup(cache_.get(), signature);
+    if (lookup.Found()) {
+      CacheElem* elem = lookup.value();
+      if (elem->IsExpired(time_now)) {
```

Comment:
Is there some garbage collection process running in the background to evict stuff from the cache? Otherwise, items could remain in there indefinitely, using up memory.

Reply:
Our cache is LRU cache.  If reaching max_size,  the oldest ones will be purged.